### ISSUE TITLE
How will we handle the many CURIEs for a concept in KG2?

### ISSUE THREAD

**Issue by edeutsch** [RTXteam/RTX#664]
Sorry if this has been discussed before, but I'm wondering how we plan to handle the many CURIEs for a single concept in KG2?

Let's say I want to pose a query about rickets. These are all the CURIEs in KG2 for rickets:
rickets = ['HP:0002748', 'DOID:10609', 'CUI:C0035579', 'CHV:0000010891', 'HPO:HP%3A0002748', 'MEDDRA:10039119', 'MEDCIN:33625', 'MEDLINEPLUS:3898', 'MESH:D012279', 'NCIT:C26878', 'NCI_NCI-GLOSS:CDR0000655123', 'NCI_NICHD:C26878', 'OMIM:MTHU006645', 'SNOMEDCT:41345002', 'EFO:0005583']

I am guessing that some edges attach to one CURIE, other edges attach to another CURIE.

To get a true picture of all the knowledge about rickets, we would need to query for them all in KG2?

If a query comes in referencing DOID:10609, will the Expand() method be obligated to expand on all synonyms and attach the results to DOID:10609 in the QueryGraph?

--side note 1: I notice HPO:HP%3A0002748 appears to be URLencode()d. Is that desirable?

--side note 2: Where's MONDO? Wasn't MONDO included in KG2? Where's MONDO:0005520?

**Comment by saramsey** [comment]
MONDO:0005520 is in KG2

![Screen Shot 2020-03-06 at 1 29 56 PM](https://user-images.githubusercontent.com/5562107/76124006-abd48000-5fae-11ea-9472-99d2ca58784c.png)

**Comment by saramsey** [comment]
previous post was from `kg2endpoint2.rtx.ai`

**Comment by saramsey** [comment]
> To get a true picture of all the knowledge about rickets, we would need to query for them all in KG2?

Currently, yes. But the plan is to provide a mechanism to simplify this. Open to ideas about how to do this.

**Comment by saramsey** [comment]
As for the node with id `HPO:HP%3A0002748`, this is from the HPO UMLS import and it is incorrect. There are 13,502 such nodes in KG2. I will put in a specific bug report about this issue. Thank you for reporting this. Should be a one-line fix.

**Comment by saramsey** [comment]
I've added #665 regarding the HPO issue.

**Comment by edeutsch** [comment]
ah, I see it. But the name is "rickets (disease)", and so a search for "rickets" will not turn up this node.

I am wondering if there should be a special rule in the MONDO import to strip out such unnecessary parenthetical add-ons.

**Comment by amykglen** [comment]
the SRI team's [NodeNormalization Service](https://github.com/TranslatorIIPrototypes/NodeNormalization) might be helpful in tackling this issue.. (for a given curie, it returns the 'preferred curie', 'equivalent identifiers', etc..)

**Comment by edeutsch** [comment]
KGNodeIndex does this as well, so I think we have all the synonym information. The question is what do we do with that information? If we get an incoming query for DOID:10609, do we just do the expand() on DOID:10609 and done and miss a lot of edges? (I assume that is what happens now?) Or do we consult KGNodeIndex for all synonyms of DOID:10609 and then perform 14 expand steps for each of the synonyms of rickets? but mapping the results for each back to DOID:10609? Maybe slow? An alternative is to coalesce information in the knowledge graph and not have 14 nodes for rickets.

**Comment by amykglen** [comment]
yes, agree with all of the above, though I think the NodeNormalizer might still have something to offer. for example, the NodeNormalizer's equivalent curies for rickets (DOID:10609) contains the MONDO node ("rickets (disease)") that KGNodeIndex misses:
 ```
{
  "DOID:10609": {
    "id": {
      "identifier": "MONDO:0005520",
      "label": "rickets (disease)"
    },
    "equivalent_identifiers": [
      {
        "identifier": "MONDO:0005520",
        "label": "rickets (disease)"
      },
      {
        "identifier": "DOID:10609"
      },
      {
        "identifier": "UMLS:C0154240"
      },
      {
        "identifier": "UMLS:C0035579"
      },
      {
        "identifier": "MESH:D012279"
      },
      {
        "identifier": "MEDDRA:10039119"
      },
      {
        "identifier": "NCIT:C26878"
      },
      {
        "identifier": "SNOMEDCT:41345002"
      },
      {
        "identifier": "SNOMEDCT:190641005"
      },
      {
        "identifier": "HP:0002748",
        "label": "Rickets"
      }
    ],
    "type": [
      "disease",
      "named_thing",
      "biological_entity",
      "disease_or_phenotypic_feature"
    ]
  }
}
```
but yes, this will be a complex issue (as mentioned on the AHM today)... just wanted to document here that the NodeNormalizer might be worth looking into while working on this.

**Comment by edeutsch** [comment]
understood, thanks.
BTW, I recall fixing this particular issue by having the most recent KGNodeIndex strip " (disease)" from MONDO nodes or adding it as a synonym. So I think that example was addressed with an up-to-date KGNodeIndex? not positive, but I meant to.

**Comment by edeutsch** [comment]
Tangent - Was this issue #664 also the reason that Noel's attempt to put his secret example disease during the demo failed? There were no edges of the correct type on the node that he tried. But there are 13 others nodes for the same thing and maybe one of those had the correct edges we wanted? dunno, just speculating.

**Comment by saramsey** [comment]
We are going to address this with weekly hackathon meetings on Fridays at 9:00 AM PDT until the hackathon/relay meeting April 27-30.

**Comment by edeutsch** [comment]
It was suggested that I summarize the issue to focus the discussion:

1) Assumption 1: We yearn to use the information in KG2 for regular operations because it contains so much more good stuff than KG1.
2) Assumption 2: Trying to support both KG1 and KG2 simultaneously leads to extra effort and is messy and confusing. We would like to consolidate all operations to KG2
3) The biggest challenge with KG2 is that in many cases it contains a dozen nodes for a concept, all with their own different edges from the different KPs.
4) In the current operation of ARAX, if a QueryGraph based on DOID:1234, edges that truly connect to the DOID:1234 node are returned, but any edges that connect to the 11 other nodes that represent the same disease are left on the table unused. This is highly detrimental to effective reasoning
5) One possible solution is to create a synonym map during KG2 building and coalesce ("identify"?) replicate nodes into a single node, along with the synonym map. The KGNodeIndex code already builds a pretty reasonable first approximation to a synonym map.
6) Another possible solution is to preserve all replicate nodes for the same concepts under different CURIEs, but adapt our Expand() and Overlay() and KGNodeIndex() and all the other systems to be appropriately handle this multiplicity. e.g., Expand() would need to find all synonyms, perform an Expand() on all synonyms and remap those back to the queried id.

Items 5 and 6 are substantially different approaches and we need to decide which we will take. Eric's opinion is that item 5 is the easier and tidier approach. It may not be 100% successfully achievable, but 97% success is good. But knowledge is messy and a perfect job for 5 is probably not achievable, and perhaps it is better to design our reasoning from the start to work in a messy world.

Basically: do we create order during the building of KG2, or do we design our reasoning to thrive on the disorder?

We need a long discussion and a documented strategy.

**Comment by dkoslicki** [comment]
Assumption 3: we will also be reaching out to other KP's for `expand` in the near future.

**Comment by edeutsch** [comment]
After substantial discussion, we came to the conclusion that doing a good job for item 5 was quite difficult and Steve would not be able to tackle this before the hackathon, AND we really need to be dealing with this complexity at the reasoning level anyway. So, the conclusion is that we are going to attempt to deal with this solely at the reasoning level as in item 6.

Initial tasks are:
- Using our Parkinson's query demo #2 as our test case,
- Eric will check/implement a synonym lookup in KGNodeIndex. Consider also using the ARS WS call
- Amy will implement an Expand on KG2 that will start with DOID:14330, determine all synonyms via KGNodeIndex, then expand a graph using all of those nodes and see what we come up with
- Then it will be up to David and Finn to see what they will do with that, and what falls to Steve

**Comment by edeutsch** [comment]
@amykglen I have just pushed to demo two new methods in KGNodeIndex.

kgNodeIndex.get_equivalent_curies('DOID:14330')
yields:
['CUI:C0030567', 'CHV:0000009319', 'ICD10:G20', 'CUI:C0865476', 'CUI:C0865475', 'CUI:C0865474', 'CUI:C1399358', 'ICD10CM:G20', 'LNC:LA27533-1', 'LNC:MTHU020807', 'MEDDRA:10061536', 'MEDCIN:32004', 'MEDLINEPLUS:85', 'MESH:D010300', 'CUI:C1521736', 'NANDA-I:03003', 'NCIT:C26845', 'NCI_KEGG:C38845', 'NCI_NCI-GLOSS:CDR0000044140', 'SNOMEDCT:49049000', 'EFO:0002508', 'MONDO:0005180', 'DOID:14330']

and

kgNodeIndex.get_equivalent_entitites('DOID:14330')
yields (trying to emulate the NodeNormalizer reference upthread, in case that seems useful for pleasing):
{
  "DOID:14330": {
    "equivalent_identifiers": [
      {
        "identifier": "CUI:C0030567",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "CHV:0000009319",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "ICD10:G20",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "CUI:C0865476",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "CUI:C0865475",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "CUI:C0865474",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "CUI:C1399358",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "ICD10CM:G20",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "LNC:LA27533-1",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "LNC:MTHU020807",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "MEDDRA:10061536",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "MEDCIN:32004",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "MEDLINEPLUS:85",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "MESH:D010300",
        "label": "PARKINSON DISEASE"
      },
      {
        "identifier": "CUI:C1521736",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "NANDA-I:03003",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "NCIT:C26845",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "NCI_KEGG:C38845",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "NCI_NCI-GLOSS:CDR0000044140",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "SNOMEDCT:49049000",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "EFO:0002508",
        "label": "PARKINSON'S DISEASE"
      },
      {
        "identifier": "MONDO:0005180",
        "label": "PARKINSON DISEASE"
      },
      {
        "identifier": "DOID:14330",
        "label": "PARKINSON'S DISEASE"
      }
    ],
    "id": {
      "identifier": "DOID:14330",
      "label": "PARKINSON'S DISEASE"
    },
    "type": [
      "disease"
    ]
  }
}

What do you think? Does this enable the next Expand() tasks for you?

It would of course be fine to use the official NodeNormalizer, but this is faster and more reliable (I hope) and only returns nodes actually in KG2.

**Comment by amykglen** [comment]
Awesome - yes, I think this does the trick! Seems like an excellent starting point for finding synonyms to me. Perhaps down the road the NodeNormalizer could become useful (like maybe if we need such functionality when working with other KPs?), or perhaps we'll want to add to these methods by also utilizing `equivalent_to` relationships in KG2, but for now this seems to work quite well.

I'll work on the expand portion. :)

**Comment by jaredroach** [comment]
Fundamentally, bullet point items #5 and #6 are the same approach. That is, we create a new kind of edge called "synonym". Whether this edge is explicitly represented as a new edge kind in KG2, or is kept in a separate table is mostly a matter of logistics, not conceptual difference. One possibility of crating a "synonym" edge under either scenario is to "weight" these edges with a confidence that the two concepts are synonyms. In either case, it would be good to keep track of provenance as to why we think two concepts are synonyms. It makes sense to have an API call that is Expand by Synonyms.

It gets tricky when doing logic on graphs that include "synonym" edges. Because generally speaking we want to consider the length of this edges to be zero, as if they are non-existent, when computing things like numbers of hops or normalized Google distance.

Also, aesthetically, most end users want to see these returned as a single node. A problem is they might be "subtly" different. E.g. the difference between a "phenotype" and a "disease". In some contexts these concepts are exactly and perfectly synonymous. In other contexts, they are not.

Things to think about for the long term - past the hackathon. These were also issues for KG1, just not as bad as KG2. And were one of my major end-user causes for dissatisfaction. For example, the major protein associated with a genetic disease might be connected to a "phenotype" node that is roughly synonymous to the OMIM entry, but not directly connected to the OMIM node.

Your solutions above a great for short term == Hackathon.

**Comment by edeutsch** [comment]
I like the idea of synonyms as edges. Certainly from the aesthetic point of view it seems pleasing. More difficult to reason on, but we should be dealing with the issue anyway. Then when a reasoner is asked to work with DOID:1234, it really should bring in all exact_synonym edges, near_synonym edges (like the phenotype/disease example), and is_child_of edges, and likewise their exact_synonyms, near_synonyms, and children and gather knowledge about all that to proceed. From the rumors I hear about neo4j, this may be painful, though.

**Comment by amykglen** [comment]
think it's safe to close this issue at this point? I suppose our solution is the node synonymizer and KG2c. :)

**Comment by dkoslicki** [comment]
Agreed and closing
